File: tests/test_rule_engine_parity.py
"""Golden/characterization parity for the Python moneyline engine.

The goldens in fixtures/moneyline_py_goldens.json were captured from the
pre-refactor apply_confidence_downgrade(). After the refactor delegates its
middle block to src.rule_engine.evaluate_moneyline(), this suite proves the
full output dict is byte-identical for every corpus case. If a golden ever
needs to change, it must be because production behavior INTENTIONALLY changed
-- never to paper over a refactor regression.
"""

import json
import os
import unittest

from src.quality_control import apply_confidence_downgrade

_FIXTURES = os.path.join(os.path.dirname(__file__), "fixtures")

with open(os.path.join(_FIXTURES, "moneyline_py_goldens.json"), encoding="utf-8") as _f:
    GOLDENS = json.load(_f)


class MoneylineParityTest(unittest.TestCase):
    def test_output_matches_captured_goldens_for_whole_corpus(self) -> None:
        import sys

        if _FIXTURES not in sys.path:
            sys.path.insert(0, os.path.dirname(__file__))
        from fixtures.moneyline_py_corpus import CORPUS

        self.assertEqual(
            set(CORPUS.keys()),
            set(GOLDENS.keys()),
            "corpus and goldens must cover the same cases",
        )

        for name, factory in CORPUS.items():
            with self.subTest(case=name):
                prediction, report = factory()
                actual = apply_confidence_downgrade(prediction, report)
                self.assertEqual(
                    actual, GOLDENS[name], f"parity mismatch on case '{name}'"
                )


if __name__ == "__main__":
    unittest.main()

File: tests/test_rule_engine_parity.js
// Golden/characterization parity for the JS moneyline engine.
//
// The goldens in fixtures/moneyline_goldens.json were captured from the
// pre-refactor applyMoneylineValueMarket(). After the refactor delegates to
// src/rule_engine.js, this suite proves the {status, reason, reasons} output is
// byte-identical for every corpus case. If a golden ever needs to change, it
// must be because production behavior INTENTIONALLY changed — never to paper
// over a refactor regression.
//
// Runs under an isolated (empty) MLB_EVOLUTION_DATA_DIR so the two
// audit-guardrail rules stay inactive and results are deterministic regardless
// of the repo's live approved_rules.json.
import assert from 'node:assert/strict';
import { mkdtempSync, rmSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import test from 'node:test';
import { readFileSync } from 'node:fs';
import { fileURLToPath } from 'node:url';

const GOLDENS = JSON.parse(
  readFileSync(fileURLToPath(new URL('./fixtures/moneyline_goldens.json', import.meta.url)), 'utf8')
);

test('moneyline engine output matches captured goldens for the whole corpus', async () => {
  const previousDir = process.env.MLB_EVOLUTION_DATA_DIR;
  const dir = mkdtempSync(join(tmpdir(), 'mlb-parity-'));
  process.env.MLB_EVOLUTION_DATA_DIR = dir;
  try {
    // Import AFTER setting the env dir; evolutionControls reads it per call so
    // an empty dir yields the advisory-only defaults (no active guardrails).
    const { applyMoneylineValueMarket } = await import('../src/mlb.js');
    const { CORPUS } = await import('./fixtures/moneyline_corpus.js');

    const names = Object.keys(CORPUS);
    assert.equal(names.length, Object.keys(GOLDENS).length, 'corpus and goldens must cover the same cases');

    for (const name of names) {
      const golden = GOLDENS[name];
      assert.ok(golden, `missing golden for corpus case ${name}`);
      const game = CORPUS[name]();
      applyMoneylineValueMarket(game);
      const actual = {
        status: game.betDecision.status,
        reason: game.betDecision.reason,
        reasons: game.betDecision.reasons
      };
      assert.deepEqual(actual, golden, `parity mismatch on case "${name}"`);
    }
  } finally {
    if (previousDir === undefined) delete process.env.MLB_EVOLUTION_DATA_DIR;
    else process.env.MLB_EVOLUTION_DATA_DIR = previousDir;
    rmSync(dir, { recursive: true, force: true });
  }
});
